edb/server/protocol/ext/auth.py

```diff
@@ -0,0 +1,282 @@
+            await handle_auth_callback(request, response)
+        else:
+            raise errors.BackendError
+    except Exception as ex:
+        if debug.flags.server:
+            markup.dump(ex)
+
+        # XXX Fix this when LSP "location" objects are implemented
+        ex_type = errors.InternalServerError
+
+        _response_error(
+            response, http.HTTPStatus.INTERNAL_SERVER_ERROR, str(ex), ex_type
+        )
+
+
+def _response_error(response, status, message, ex_type):
+    err_dct = {
+        "message": message,
+        "type": str(ex_type.__name__),
+        "code": ex_type.get_code(),
+    }
+
+    response.body = json.dumps({"error": err_dct}).encode()
+    response.status = status
+    response.close_connection = True
+
+
+# OAuth 2 Provider for GitHub
+class GitHubProvider(BaseProvider):
+    def __init__(self, client_id, client_secret):
+        super().__init__("github", client_id, client_secret)
+
+    def get_code_url(
+        self, state: str, redirect_uri: str, scope: str = "read:user"
+    ) -> str:
+        params = {
+            "client_id": self.client_id,
+            "scope": scope,
+            "state": state,
+            "redirect_uri": redirect_uri,
+        }
+        encoded = urllib.parse.urlencode(params)
+        return f"https://github.com/login/oauth/authorize?{encoded}"
+
+    async def exchange_access_token(self, code, state):
+        # Check state value
+        # TODO: Look up state value from FlowState object
+        # flow_state = await db.get_flow_state(state, "github")
+        # if flow_state is None:
+        #    raise errors.UnauthorizedError("invalid state value")
+        data = {
+            'grant_type': 'authorization_code',
+            'code': code,
+            'state': state,
+            "redirect_uri": "http://localhost:8080/auth/callback",  # TODO: Get correct db-namespaced URL
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+        }
+
+        headers = {'Content-Type': 'application/json'}
+
+        async with httpx.AsyncClient() as client:
+            resp = await client.post(
+                "https://github.com/login/oauth/access_token",
+                json=data,
+                headers=headers,
+            )
+            token = resp.json()['access_token']
+
+            return token
+
+
+providers = {
+    "github": GitHubProvider,
+}
+
+
+def _get_provider(name, client_id, client_secret):
+    provider_class = providers.get(name)
+    if provider_class is None:
+        raise errors.InternalServerError(f"unknown provider: {name}")
+    return provider_class(client_id, client_secret)
+
+
+def _get_auth_signing_key(db_config: SettingsMap):
+    auth_signing_key = _maybe_get_config(db_config, "xxx_auth_signing_key")
+    if auth_signing_key is None:
+        raise errors.InternalServerError(
+            "No auth signing key configured: Please set `cfg::Config.xxx_auth_signing_key`"
```

Comment:
nit: Our code style dictates max 79 characters per line
```suggestion
            "No auth signing key configured: "
            "Please set `cfg::Config.xxx_auth_signing_key`"
```